working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,913 @@
+      )
+    ),
+  );
+}
+```
+
+### Type inference
+
+This proposal is mostly about readability, but that isn't the only benefit. By
+turning imperative, mutating code into declarative expressions inside the
+collection literal, type inference becomes more effective. Both upwards and
+downwards inference has more code to chew on.
+
+Consider:
+
+```dart
+Widget build(BuildContext context) {
+  var buttons = <Widget>[];
+
+  if (isAndroid) {
+    buttons.add(IconButton(icon: Icon(Icons.search)));
+  }
+
+  buttons.add(IconButton(icon: Icon(Icons.menu)));
+
+  return Row(children: buttons);
+}
+```
+
+Note that explicit type annotation on the list literal. That's needed because we
+don't know any of its elements at creation time. By moving that `if` inside the
+list, we can use the elements to infer the list's type:
+
+```dart
+Widget build(BuildContext context) {
+  var buttons = [
+    if (isAndroid) IconButton(icon: Icon(Icons.search)),
+    IconButton(icon: Icon(Icons.menu))
+  ];
+
+  return Row(children: buttons);
+}
+```
+
+In many cases, this upwards inference infers the type that you want, and being
+able to move more of the list's contents inside the literal improves that. In
+this case, though, the inferred type is a little more precise than desired.
+Fortunately, downwards inference is improved too. If the code is fully
+refactored to:
+
+```dart
+Widget build(BuildContext context) {
+  return Row(children: [
+    if (isAndroid) IconButton(icon: Icon(Icons.search)),
+    IconButton(icon: Icon(Icons.menu))
+  ]);
+}
+```
+
+Now, the fact that Row's `children` parameter has type `List<Widget>` causes
+that to be the inferred type of the list. We're able to do this because the
+entire list creation is now a single expression so it can be moved right into
+the constructor call for `Row()`.
+
+## Syntax
+
+We extend the list grammar to allow *control flow elements* in addition to
+regular elements:
+
+```
+listLiteral:
+  const? typeArguments? '[' listElementList? ']'
+  ;
+
+listElementList:
+  listElement ( ',' listElement )* ','?
+  ;
+
+listElement:
+  expression |
+  'if' '(' expression ')' listElement ( 'else' listElement )? |
+  'for' '(' forLoopParts ')' listElement
+  ;
+```
+
+Instead of `expressionList`, this uses a new `listElementList` rule since
+`expressionList` is used elsewhere in the grammar like argument lists where
+control flow isn't allowed.
+
+Each element in a list can be one of a few things:
+
+* A normal expression.
+* An `if` element.
+* A `for` element.
+
+The body of `if` and `for` elements use `listElement`, not `expression`, which
+allows nesting.
+
+The changes for map literals are similar:
+
+```
+mapLiteral:
+  const? typeArguments? '{' mapLiteralEntryList? '}' ;
+
+mapLiteralEntryList:
+  mapLiteralEntry ( ',' mapLiteralEntry )* ','?
+  ;
+
+mapLiteralEntry:
+  expression ':' expression |
+  'if' '(' expression ')' mapLiteralEntry ( 'else' mapLiteralEntry )? |
+  'for' '(' forLoopParts ')' mapLiteralEntry
```

Comment:
Consider allowing `await for` in an async function.
We do allow `await` in the expressions, and the user can get the same effect as:
```dart
await for (var s in stream) something(s)
```
as 
```dart
for (var s in await stream.toList()) something(s)
```
it's just less efficient and direct.

Reply:
OK, took a stab at adding this.